File: client/windows/GhostClient/Core/Stat.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;

namespace Ghost.Core
{
    /// <summary>
    /// This class represents a numerical stat that can have temporary buffs applied to it's value to alter the value returned.
    /// The actual value of this stat is called the TrueValue, and the computed value is the Value. The TrueValue of a stat should not be changed
    /// unless it's considered a perminate change, as this will result in unexpected results when many Buffs are applied.
    /// </summary>
    public sealed class Stat : IEquatable<Stat>, ICloneable
    {
        [Flags]
        public enum BuffType
        {
            Addition = 1,
            Subtraction = 2,
            Percentage = 4,
            AddPercent = Addition | Percentage,
            SubtractPercent = Subtraction | Percentage
        }

        public struct Buff
        {
            public BuffType Type;
            public double BuffValue;
            public string Name;
        }

        private bool _dirty = true;
        private double _trueValue;
        private double _cachedValue;
        private readonly object _valueLock = new object();
        private readonly List<Buff> _buffs = new List<Buff>();

        public IEnumerable<Buff> Buffs
        {
            get { return new ReadOnlyCollection<Buff>(_buffs); }
        }

        public double TrueValue
        {
            get { return _trueValue; }
            set
            {
                _trueValue = value;
                _dirty = true;
            }
        }

        public double Value
        {
            get
            {
                lock (_valueLock)
                {
                    if (!_dirty) return _cachedValue;

                    _cachedValue = TrueValue;
                    foreach (Buff buff in _buffs)
                    {
                        bool percent = (buff.Type & BuffType.Percentage) != 0;
                        if ((buff.Type & BuffType.Subtraction) != 0)
                        {
                            if (percent)
                            {
                                _cachedValue -= _cachedValue * (buff.BuffValue / 100.0);
                            }
                            else
                            {
                                _cachedValue -= buff.BuffValue;
                            }
                        }
                        else
                        {
                            if (percent)
                            {
                                _cachedValue += _cachedValue * (buff.BuffValue / 100.0);
                            }
                            else
                            {
                                _cachedValue += buff.BuffValue;
                            }
                        }
                    }

                    _dirty = false;
                }

                return _cachedValue;
            }
        }

        public Stat(Stat stat)
        {
            TrueValue = stat.TrueValue;
            _buffs = stat._buffs;
        }

        public Stat(double trueValue)
        {
            TrueValue = trueValue;
        }

        public Stat()
            : this(0.0)
        {
        }

        #region Operator Override
        public static implicit operator Stat(double value)
        {
            return new Stat(value);
        }

        public static implicit operator Stat(float value)
        {
            return new Stat(value);
        }

        public static implicit operator Stat(int value)
        {
            return new Stat(value);
        }

        public static Stat operator +(Stat stat, double value)
        {
            stat.TrueValue += value;
            return stat;
        }

        public static Stat operator -(Stat stat, double value)
        {
            stat.TrueValue -= value;
            return stat;
        }

        public static Stat operator *(Stat stat, double value)
        {
            stat.TrueValue *= value;
            return stat;
        }

        public static Stat operator /(Stat stat, double value)
        {
            stat.TrueValue /= value;
            return stat;
        }

        public static Stat operator +(Stat stat1, Stat stat2)
        {
            var toReturn = new Stat { TrueValue = stat1.TrueValue + stat2.TrueValue };
            toReturn._buffs.AddRange(stat1._buffs.Union(stat2._buffs));

            return toReturn;
        }

        public static Stat operator +(Stat stat, float value)
        {
            stat.TrueValue += value;
            return stat;
        }

        public static Stat operator -(Stat stat, float value)
        {
            stat.TrueValue -= value;
            return stat;
        }

        public static Stat operator *(Stat stat, float value)
        {
            stat.TrueValue *= value;
            return stat;
        }

        public static Stat operator /(Stat stat, float value)
        {
            stat.TrueValue /= value;
            return stat;
        }
        public static Stat operator +(Stat stat, int value)
        {
            stat.TrueValue += value;
            return stat;
        }

        public static Stat operator -(Stat stat, int value)
        {
            stat.TrueValue -= value;
            return stat;
        }

        public static Stat operator *(Stat stat, int value)
        {
            stat.TrueValue *= value;
            return stat;
        }

        public static Stat operator /(Stat stat, int value)
        {
            stat.TrueValue /= value;
            return stat;
        }

        public static bool operator ==(Stat stat, int value)
        {
            return stat != null && stat.Value == value;
        }

        public static bool operator !=(Stat stat, int value)
        {
            return stat == null || stat.Value != value;
        }

        public static bool operator >(Stat stat, int value)
        {
            return stat != null && stat.Value > value;
        }

        public static bool operator <(Stat stat, int value)
        {
            return stat == null || stat.Value < value;
        }

        public static bool operator >=(Stat stat, int value)
        {
            return stat != null && stat.Value >= value;
        }

        public static bool operator <=(Stat stat, int value)
        {
            return stat == null || stat.Value <= value;
        }

        public static bool operator ==(Stat stat, double value)
        {
            return stat != null && stat.Value == value;
        }

        public static bool operator !=(Stat stat, double value)
        {
            return stat == null || stat.Value != value;
        }

        public static bool operator >(Stat stat, double value)
        {
            return stat != null && stat.Value > value;
        }

        public static bool operator <(Stat stat, double value)
        {
            return stat == null || stat.Value < value;
        }

        public static bool operator >=(Stat stat, double value)
        {
            return stat != null && stat.Value >= value;
        }

        public static bool operator <=(Stat stat, double value)
        {
            return stat == null || stat.Value <= value;
        }

        public static bool operator ==(Stat stat, float value)
        {
            return stat != null && stat.Value == value;
        }

        public static bool operator !=(Stat stat, float value)
        {
            return stat == null || stat.Value != value;
        }

        public static bool operator >(Stat stat, float value)
        {
            return stat != null && stat.Value > value;
        }

        public static bool operator <(Stat stat, float value)
        {
            return stat == null || stat.Value < value;
        }

        public static bool operator >=(Stat stat, float value)
        {
            return stat != null && stat.Value >= value;
        }

        public static bool operator <=(Stat stat, float value)
        {
            return stat == null || stat.Value <= value;
        }



        public bool Equals(Stat other)
        {
            if (ReferenceEquals(null, other)) return false;
            if (ReferenceEquals(this, other)) return true;
            return Equals(_buffs, other._buffs) && TrueValue.Equals(other.TrueValue);
        }

        public override int GetHashCode()
        {
            unchecked
            {
                return ((_buffs != null ? _buffs.GetHashCode() : 0) * 397) ^ TrueValue.GetHashCode();
            }
        }

        public object Clone()
        {
            return new Stat(this);
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            return obj is Stat && Equals((Stat)obj);
        }

        public override string ToString()
        {
            return Value.ToString();
        }

        #endregion

        public Buff AddBuff(BuffType type, double buffValue, string name, bool stack = true)
        {
            _dirty = true;

            if (!stack)
            {
                Buff temp = _buffs.FirstOrDefault(b => b.Name == name);
                if (!EqualityComparer<Buff>.Default.Equals(temp, default(Buff)))
                {
                    temp.Type = type;
                    temp.BuffValue = buffValue;
                    return temp;
                }
            }

            var buff = new Buff { Type = type, BuffValue = buffValue, Name = name };
            _buffs.Add(buff);
            return buff;
        }

        public Buff AddBuff(Buff buff)
        {
            if (_buffs.Contains(buff)) return buff;
            _dirty = true;
            _buffs.Add(buff);
            return buff;
        }

        public void RemoveBuff(Buff buff)
        {
            if (!_buffs.Contains(buff)) return;
            _dirty = true;
            _buffs.Remove(buff);
        }

        public void RemoveBuff(string name)
        {
            _dirty = true;

            _buffs.RemoveAll(b => b.Name == name);
        }

        public void ClearBuffs()
        {
            _dirty = true;

            _buffs.Clear();
        }
    }
}

File: server/api/src/main/java/me/eddiep/ghost/game/match/stats/Stat.java
package me.eddiep.ghost.game.match.stats;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * A Stat represents a numerical stat that can have temporary buffs applied to its value to alter the value returned.
 * The true value of a stat should not be changed unless it's considered a permanent change, as this will result in
 * unexpected results when many Buffs are applied.
 */

public final class Stat {

    private final String id;

    private boolean dirty = true;
    private double trueValue;
    private double cachedValue;

    private final Object valueLock = new Object();

    private final List<Buff> buffs = new ArrayList<>();

    public Stat(String id, Stat stat) {
        this.id = id;

        this.trueValue = stat.trueValue;
        this.buffs.addAll(stat.buffs);
    }

    public Stat(String id, double value) {
        this.id = id;

        this.trueValue = value;
    }

    public Stat(String id) {
        this(id, 0.0);
    }

    /**
     * Returns the network ID of this stat.<br />
     * This returns the ID without any checks. An ID should consist of exactly 4 bytes. When sending the stat update
     * packet, a shorter ID will be left-padded with spaces, but a longer one will throw an exception.
     */
    public String getId() {
        return id;
    }

    /**
     * Returns an unmodifiable list of the buffs applied to this stat.
     */
    public List<Buff> getBuffs() {
        return Collections.unmodifiableList(buffs);
    }

    /**
     * Adds a new buff to this stat.<br />
     *
     * Trying to add an unstackable buff that already exists will overwrite the value of the previous buff instead of
     * adding another one.
     *
     * @param name The name of the buff. Used for unstackable buffs.
     * @param type The type of the buff.
     * @param value The value of the buff.
     * @param stack Whether the buff is stackable or not. If a buff is stackable, then more than one can be applied and
     *              all will have an effect on the base value. If a buff is not stackable, then only one can be applied
     *              to a stat at a time.
     */
    public Buff addBuff(String name, BuffType type, double value, boolean stack) {
        dirty = true;

        if (!stack) {
            for (Buff buff: buffs) { //Java 7 makes me cry irl.
                if (buff.getName().equals(name)) {
                    buff.setType(type);
                    buff.setValue(value);

                    return buff;
                }
            }
        }

        Buff buff = new SimpleBuff(name, type, value);
        addBuff(buff);

        return buff;
    }

    /**
     * Adds a new buff to this stat that lasts for x seconds! <br />
     *
     * Trying to add an unstackable buff that already exists will overwrite the value of the previous buff instead of
     * adding another one.
     *
     * @param name The name of the buff. Used for unstackable buffs.
     * @param type The type of the buff.
     * @param value The value of the buff.
     * @param stack Whether the buff is stackable or not. If a buff is stackable, then more than one can be applied and
     *              all will have an effect on the base value. If a buff is not stackable, then only one can be applied
     *              to a stat at a time.
     * @param duration The duration, in seconds that indicates how long this buff lasts
     */
    public Buff addTimedBuff(String name, BuffType type, double value, boolean stack, double duration) {
        dirty = true;

        if (!stack) {
            for (Buff buff : buffs) { //Java 7 makes me cry irl.
                if (buff.getName().equals(name)) {
                    buff.setType(type);
                    buff.setValue(value);

                    return buff;
                }
            }
        }

        Buff buff = new TimedBuff(name, type, value, (long)(1000 * duration));
        addBuff(buff);

        return buff;
    }

    /**
     * Adds a stackable buff to this stat that expires in x seconds. Equivalent to {@code addBuff(name, type, value, true);}
     *
     * @param duration The duration, in seconds that indicates how long this buff lasts
     *
     * @see #addBuff(String, BuffType, double, boolean)
     */
    public Buff addTimedBuff(String name, BuffType type, double value, double duration) {
        return addTimedBuff(name, type, value, true, duration);
    }

    /**
     * Adds a stackable buff to this stat. Equivalent to {@code addBuff(name, type, value, true);}
     *
     * @see #addBuff(String, BuffType, double, boolean)
     */
    public Buff addBuff(String name, BuffType type, double value) {
        return addBuff(name, type, value, true);
    }

    public void removeBuff(SimpleBuff buff) {
        dirty = buffs.remove(buff);
    }

    public void removeBuff(String name) {
        for (Buff buff: buffs) { //Why Java7 why.
            if (buff.getName().equals(name)) {
                buffs.remove(buff);
                dirty = true;
                break;
            }
        }
    }

    public void addBuff(Buff buff) {
        buff.apply();
        buffs.add(buff);
    }

    /**
     * Removes all of the applied buffs.
     */
    public void removeBuffs() {
        buffs.clear();
        dirty = true;
    }

    /**
     * Returns the true value of this stat, without any buffs applied to it. Use the {@link #getValue()} method to get
     * the effective value of the buff.
     */
    public double getTrueValue() {
        return trueValue;
    }

    /**
     * Changes the true value of this stat. Unless the change is permanent, consider using {@link SimpleBuff buffs} instead.
     *
     * @param trueValue The new value of the buff.
     */
    public void setTrueValue(double trueValue) {
        this.trueValue = trueValue;
        dirty = true;
    }

    /**
     * Gets the value of this stat, with all the buffs applied to it. a cached value is used unless a modification has
     * occurred.
     */
    public double getValue() {
        synchronized (valueLock) {
            if (!dirty) {
                return cachedValue;
            }

            cachedValue = trueValue;

            for (Buff buff : buffs) {
                boolean percent = buff.hasBuff(BuffType.Percentage);

                if (buff.hasBuff(BuffType.Subtraction)) {
                    if (percent) {
                        cachedValue -= (buff.getValue() / 100.0) * cachedValue;
                    } else {
                        cachedValue -= buff.getValue();
                    }
                } else if (buff.hasBuff(BuffType.Addition)) {
                    if (percent) {
                        cachedValue += (buff.getValue() / 100.0) * cachedValue;
                    } else {
                        cachedValue += buff.getValue();
                    }
                }
            }

            dirty = false;
        }

        return cachedValue;
    }

    public boolean hasBuff(String name) {
        for (Buff buff : buffs) {
            if (buff.getName().equals(name))
                return true;
        }
        return false;
    }

    public void tick() {
        for (Buff buff : buffs) {
            buff.tick(this);
        }
    }
}
